share/CMakeJS.cmake

```diff
@@ -0,0 +1,640 @@
+#[=============================================================================[
+Check whether we have already been included (borrowed from CMakeRC)
+]=============================================================================]#
+# Hypothetical CMakeJS version number...
+set(_version 8.0.0)
+
+cmake_minimum_required(VERSION 3.15)
+cmake_policy(VERSION 3.15)
+include(CMakeParseArguments)
+
+if(COMMAND cmakejs_napi_addon_add_sources)
+    if(NOT DEFINED _CMAKEJS_VERSION OR NOT (_version STREQUAL _CMAKEJS_VERSION))
+        message(WARNING "More than one CMakeJS version has been included in this project.")
+    endif()
+    # CMakeJS has already been included! Don't do anything
+    return()
+endif()
+
+set(_CMAKEJS_VERSION "${_version}" CACHE INTERNAL "CMakeJS version. Used for checking for conflicts")
+
+set(_CMAKEJS_SCRIPT "${CMAKE_CURRENT_LIST_FILE}" CACHE INTERNAL "Path to 'CMakeJS.cmake' script")
+
+# Default build output directory, if not specified with '-DCMAKEJS_BINARY_DIR:PATH=/some/dir'
+if(NOT DEFINED CMAKEJS_BINARY_DIR)
+    set(CMAKEJS_BINARY_DIR "${CMAKE_BINARY_DIR}")
+endif()
+
+#[=============================================================================[
+Internal helper (borrowed from CMakeRC).
+]=============================================================================]#
+function(_cmakejs_normalize_path var)
+    set(path "${${var}}")
+    file(TO_CMAKE_PATH "${path}" path)
+    while(path MATCHES "//")
+        string(REPLACE "//" "/" path "${path}")
+    endwhile()
+    string(REGEX REPLACE "/+$" "" path "${path}")
+    set("${var}" "${path}" PARENT_SCOPE)
+endfunction()
+
+#[=======================================================================[
+FindCMakeJs.cmake
+--------
+
+Find the native CMakeJs includes, source, and library
+
+(This codeblock typically belongs in a file named 'FindCMakeJS.cmake' for
+distribution...)
+
+This module defines
+
+::
+
+  CMAKE_JS_INC, where to find node.h, etc.
+  CMAKE_JS_LIB, the libraries required to use CMakeJs.
+  CMAKE_JS_SRC, where to find required *.cpp files, if any,
+  CMAKE_JS_EXECUTABLE, the cmake-js binary (global)
+  CMAKE_JS_NPM_PACKAGE, the cmake-js binary (local)
+
+]=======================================================================]#
+
+# CMAKE_JS_VERSION is defined on all platforms when calling from cmake-js.
+# By checking whether this var is pre-defined, we can determine if we are
+# running from an npm script (via cmake-js), or from CMake directly...
+
+if (NOT DEFINED CMAKE_JS_VERSION) # TODO: Unfortunately, this var is a little too close to 'CMAKE_VERSION' for comfort... Kitware may need it. Should be moved into 'CMAKEJS_VERSION'!
+
+    # ...and if we're calling from CMake directly, we need to set up some vars
+    # that our build step depends on (these are predefined when calling via npm/cmake-js).
+    if(VERBOSE)
+        message(STATUS "CMake Calling...")
+    endif()
+
+    # Check for cmake-js installations
+    find_program(CMAKE_JS_EXECUTABLE
```

Comment:
Does this need a `find_program`? The path will always be '../bin/cmake-js` relative to this `CMakeJS.cmake` file, so it seems to me like it could be hardcoded. (and that would avoid potential issues like it finding a different version of cmake-js to what this file is written for)

If we do need to search for this, then I think it would make sense to support doing `-DCMAKE_JS_EXECUTABLE=.....`, so that if this fails (It certainly will on my machine, I never put cmake-js in the path) then the location can be 'forced'.

Reply:
oh right, you are expecting it to be installed globally. which I have never done.
The 'local' version is then being searched for below. I havent looked far enough to see why both are needed yet

Reply:
Good questions! It raises the point that it's really hard to predict all usage cases.

The main objective of *all* the other code before the several ```add_library()```'s is just to make sure those libs contain the specified header files, defs, etc. Essentially, if you do leverage all those header files and stuff some other way instead (and manage the 'building-with-native-cmake' thing that people seem to want), then I don't see any issue. The result is more important than the means, within the boundaries of convention.

I admit I personally mostly just use ```cmake-js install```, ```rebuild```, and ```reconfigure``` commands, little else, most of the time. I had thought that the global install was something cmake-js requires. It always "just worked" for me; no doubt there are lots of scenarios and ways that people are using cmake-js (or plan to) that I hadn't imagined. But if the API is really designed well, it should help those people, too, hopefully.

I am not yet sure if this suggestion is logical, but I'm wondering whether cmake-js actually needs to do that global install step at all? I note that there is some version sync workload, keeping cmake-js in sync with NodeJS versions. I am curious if there is some opportunity in here to not just... take what one needs from the running node instance's installation folder? I gather that the global install is a psuedo equivalent to ```sudo apt install libnode-dev``` on Debian platforms. I'm not really 100% sure - I've kept my input solely to the CMake API for a reason! - but I believe there may be scope for improving the handling of cmake-js' dependency workload, by leaning a bit more into CMake. 

Just thoughts for your side.

Reply:
I have another comment below which suggests a change to be made here.

We don't need to worry about the 'full' headers for now, lets focus on getting this done for node-api then other projects can be solved in a second iteration of this.